Add Tensor::is_cpu, genericize TensorIterator

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53973 Add some trivial meta functions for commonly used inplace operations
* #54127 Fix inaccurate dispatch tables
* **#54079 Add Tensor::is_cpu, genericize TensorIterator**
* #54034 Refactor tensor_new.cpp to use TensorOptions instead of DispatchKey

Fixes https://github.com/pytorch/pytorch/issues/53815

Instead of testing if something is CUDA, we instead test if something
is not CPU.  This in the general theming of "Don't be so darn CUDA
centric".

Intruigingly, we didn't have a is_cpu() method on Tensor.  Which seems
like a big oversight and one of the reasons how we ended up in this
mess.  So in it goes.  Maybe we should also get this for Python bindings
as well (but in that case, should probably look into redoing all of the
is_X bindings so they aren't done manually).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27109507](https://our.internmc.facebook.com/intern/diff/D27109507)

ezyang committed:
Add Tensor::is_cpu, genericize TensorIterator

Fixes https://github.com/pytorch/pytorch/issues/53815

Instead of testing if something is CUDA, we instead test if something
is not CPU.  This in the general theming of "Don't be so darn CUDA
centric".

Intruigingly, we didn't have a is_cpu() method on Tensor.  Which seems
like a big oversight and one of the reasons how we ended up in this
mess.  So in it goes.  Maybe we should also get this for Python bindings
as well (but in that case, should probably look into redoing all of the
is_X bindings so they aren't done manually).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ngimel commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -324,12 +324,12 @@ void TensorIteratorBase::compute_types(const TensorIteratorConfig& config) {
-      if ((common_device.is_cuda() || common_device.is_xpu()) &&
+      if (!common_device.is_cpu() &&
           config.allow_cpu_scalars_ && !op.is_output && op.tensor.dim() == 0 &&
           op.tensor.device().is_cpu()) {
```
can this become op.tensor.is_cpu()?

ngimel commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -857,7 +857,7 @@ TensorIterator TensorIterator::reduce_op(Tensor& out, const Tensor& a) {
   TORCH_INTERNAL_ASSERT(out1.defined());
   TORCH_INTERNAL_ASSERT(out2.defined());
-  TORCH_CHECK((!a.is_cuda() && !out1.is_cuda() && !out2.is_cuda()) || (a.device() == out1.device() && out1.device() == out2.device()),
+  TORCH_CHECK((a.is_cpu() && out1.is_cpu() && out2.is_cpu()) || (a.device() == out1.device() && out1.device() == out2.device()),
```
looks like this is not needed at all, because this will be handled in `build` since `check_all_same_device_ = true`

albanD reviewed (commented):
Looks good with natalia's updates
Is there any plan to make these available in python? We have a couple that are bound but not all.

ezyang committed:
Update on "Add Tensor::is_cpu, genericize TensorIterator"


Fixes https://github.com/pytorch/pytorch/issues/53815

Instead of testing if something is CUDA, we instead test if something
is not CPU.  This in the general theming of "Don't be so darn CUDA
centric".

Intruigingly, we didn't have a is_cpu() method on Tensor.  Which seems
like a big oversight and one of the reasons how we ended up in this
mess.  So in it goes.  Maybe we should also get this for Python bindings
as well (but in that case, should probably look into redoing all of the
is_X bindings so they aren't done manually).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
> Is there any plan to make these available in python? We have a couple that are bound but not all.

I think it would make sense. Not this PR.

ezyang committed:
Update on "Add Tensor::is_cpu, genericize TensorIterator"


Fixes https://github.com/pytorch/pytorch/issues/53815

Instead of testing if something is CUDA, we instead test if something
is not CPU.  This in the general theming of "Don't be so darn CUDA
centric".

Intruigingly, we didn't have a is_cpu() method on Tensor.  Which seems
like a big oversight and one of the reasons how we ended up in this
mess.  So in it goes.  Maybe we should also get this for Python bindings
as well (but in that case, should probably look into redoing all of the
is_X bindings so they aren't done manually).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27109507](https://our.internmc.facebook.com/intern/diff/D27109507)

[ghstack-poisoned]

ezyang committed:
Update on "Add Tensor::is_cpu, genericize TensorIterator"


Fixes https://github.com/pytorch/pytorch/issues/53815

Instead of testing if something is CUDA, we instead test if something
is not CPU.  This in the general theming of "Don't be so darn CUDA
centric".

Intruigingly, we didn't have a is_cpu() method on Tensor.  Which seems
like a big oversight and one of the reasons how we ended up in this
mess.  So in it goes.  Maybe we should also get this for Python bindings
as well (but in that case, should probably look into redoing all of the
is_X bindings so they aren't done manually).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27109507](https://our.internmc.facebook.com/intern/diff/D27109507)

[ghstack-poisoned]

ezyang committed:
Update on "Add Tensor::is_cpu, genericize TensorIterator"


Fixes https://github.com/pytorch/pytorch/issues/53815

Instead of testing if something is CUDA, we instead test if something
is not CPU.  This in the general theming of "Don't be so darn CUDA
centric".

Intruigingly, we didn't have a is_cpu() method on Tensor.  Which seems
like a big oversight and one of the reasons how we ended up in this
mess.  So in it goes.  Maybe we should also get this for Python bindings
as well (but in that case, should probably look into redoing all of the
is_X bindings so they aren't done manually).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27109507](https://our.internmc.facebook.com/intern/diff/D27109507)

[ghstack-poisoned]
